proposals/function-pointers.md

```diff
@@ -36,15 +36,25 @@ of a `delegate*` will use `calli` where invocation of a `delegate` will use `cal
 Syntactically though invocation is identical for both constructs.
 
 The ECMA-335 definition of method pointers includes the calling convention as part of the type signature (section 7.1).
-The default calling convention will be `managed`. Alternate forms can be specified by adding the appropriate modifier
-after the `delegate*` syntax: `managed`, `cdecl`, `stdcall`, `thiscall`, or `unmanaged`. Example:
+The default calling convention will be `managed`. Unmanaged calling conventions can by specified by putting an `unmanaged`
+keyword afer the `delegate*` syntax, which will use the runtime platform default. Specific unmanaged conventions can then
+be specified in brackets to the `unmanaged` keyword by specifying any type starting with `CallConv` in the
+`System.Runtime.CompilerServices` namespace. These types must come from the program's core library, and the set of valid
+combinations is platform-dependent.
 
 ``` csharp
+//This method has a managed calling convention. This is the same as leaving the managed keyword off.
+delegate* managed<int, int>;
+
+// This method will be invoked using whatever the default unmanaged calling convention on the runtime
+// platform is. This is platform and architecture dependent and is determined by the CLR at runtime.
+delegate* unmanged<int, int>;
+
 // This method will be invoked using the cdecl calling convention
-delegate* cdecl<int, int>;
+delegate* unmanged[CallConvCdecl] <int, int>;
 
-// This method will be invoked using the stdcall calling convention
-delegate* stdcall<int, int>;
+// This method will be invoked using the stdcall calling convention, and suppresses GC transition
+delegate* unmanaged[CallConvStdCall, CallConvSuppressGCTransition] <int, int>;
```

Comment:
It looks like there is a distinction between the calling convention (like `StdCall`) and calling convention "modifier" (like `SuppressGCTransition`), but this proposal ignores that.

Would it make sense to somehow perform basic checking based on that, so that e.g. the nonsensical `unmanaged[CallConvCdecl, CallConvStdCall]` did not compile? Even if the compiler didn't understand exactly which combinations of calling conventions and their modifiers are allowed, I think such basic checking could still be worthwhile.

Reply:
It's not readily possible to know what combinations are valid and what are not or what will/wont be in the future. For example `thiscall` can be combined with basically any other convention is in some cases (`thiscall`+`stdcall` is the default for x86 COM, for example).

It's easier to just allow any combination and let the runtime decide.

Reply:
>Would it make sense to somehow perform basic checking based on that

@svick This concern was raised during design. The reason for not checking is an attempt to avoid having to update the compiler whenever a new combination is supported by the runtime. As @tannergooding mentions there are some combinations that could be described as a combination of multiple `CallConv*` types. Letting the implementing runtime decide what it supports avoids encoding a support matrix in more than one location (e.g. Roslyn and runtime) and lessens the time to market for consumption of those combinations.

The counter argument here is that deferring the error until run time is less than ideal for confidence in a compiled assembly. We can all appreciate that concern. The benefit of quicker adoption of desired combinations in an already advanced interop scenario outweighed the cost of degraded compile time validation. This is why the decision was made.

Reply:
@AaronRobinsonMSFT @tannergooding I get that hardcoding the full support matrix into the compiler is undesirable.

What I was proposing is that there would be two kinds of `CallConv`: "base" (like `Cdecl` or `StdCall`) and "modifier" (like `Thiscall` or `SuppressGCTransition`). An `unmanaged` function pointer could then have:

* Zero or one base `CallConv`.
* Zero or more modifier `CallConv`.

I think that this way, all valid combinations will still be allowed while many invalid combination will be disallowed. Or am I wrong?

Though thinking about it some more, I think it would actually be possible to encode the support matrix into the metadata of the corelib reference assembly: each base `CallConv` could have an attribute that specified the modifier `CallConv`s that it allows, or even the combination of modifier `CallConv` it allows.

Using this approach, the compiler can know the support matrix, even though altering the support matrix does not involve updating the compiler.

Reply:
That would still require updating the compiler any time a new convention is released. Having an analyzer that flags potentially bad or unsupported combinations is likely a much better alternative as it is more trivially updated, shipped, and ignored (if desired).

The support might also differ across various runtime implementations. That is, while CoreCLR might support `X, Y, and Z`; you might have another runtime like `Unity` or a `CIL to LLVM` transpiler which supports a different set (or different combinations).

Reply:
@tannergooding

> That would still require updating the compiler any time a new convention is released. 

It wouldn't, but I forgot to explain that: the compiler would differentiate between base and modifier `CallConv` based on metadata. That could be represented by a base class of the `CallConv` class (e.g. `public class CallConvThiscall :  ModifierCallConv { … }`), or an attribute applied to the `CallConv` class.

> The support might also differ across various runtime implementations.

If they use the same corlib reference assembly, yes, that would be a problem.

Reply:
> That would still require updating the compiler any time a new convention is released.

I've thought a lot about this and I keep asking myself the following: is that so bad? The compiler and runtime release as a pair now. Updating the calling convention is a small amount of work, or would be if we designed for this future. So is it really that bad if we said that every time the runtime adds a new calling convention that we have to change the compiler to support it?

Reply:
There are many runtime failure modes for interop that have complex conditions and that the compiler does not check for today. One example from many - the compiler does not check whether `MarshalAs` makes sense for given argument type in regular PInvokes. We could duplicate the runtime checks for all these in the compiler, but the value of it seems very low. I do not think that the support matrix for the unmanaged calling convention modifiers should be on different plan from the rest of interop.